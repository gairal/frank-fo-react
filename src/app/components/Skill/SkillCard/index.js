import skillCard from './SkillCard';

export default skillCard;
