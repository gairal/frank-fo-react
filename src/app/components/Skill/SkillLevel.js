import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { CircularProgress } from '@material-ui/core';

export default class SkillLevel extends Component {
  static propTypes = {
    level: PropTypes.number.isRequired,
  };

  constructor(props) {
    super(props);

    this.state = {
      progress: 0,
    };

    this.interval = null;
  }

  componentDidMount() {
    this.animateLevel();
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  updateLevel(resolve) {
    const { level } = this.props;
    const { progress } = this.state;
    if (progress >= level) {
      clearInterval(this.interval);
      resolve(progress);
    } else {
      this.setState(prevState => ({
        progress: prevState.progress + 5,
      }));
    }
  }

  animateLevel() {
    return new Promise(resolve => {
      this.interval = setInterval(() => this.updateLevel(resolve), 5);
    });
  }

  render() {
    const { progress } = this.state;
    return <CircularProgress variant="static" value={progress} thickness={5} />;
  }
}
