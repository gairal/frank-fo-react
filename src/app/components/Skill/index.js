import skill from './Skill';

export default skill;
