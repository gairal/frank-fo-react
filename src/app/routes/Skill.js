import component from '@/components/Skill';
import AbstractRoute from './AbstractRoute';

class Skill extends AbstractRoute {
  constructor(store) {
    super(store, 'skills', 'build', 'categories');
    this.component = component;

    this.ACTIONS = {
      FETCH_SUCCESS: 'SKILL_FETCH_SUCCESS',
      FETCH_FAILURE: 'SKILL_FETCH_FAILURE',
    };
    this.init();
  }

  // actions
  static success(json) {
    return {
      type: 'SKILL_FETCH_SUCCESS',
      payload: {
        json,
      },
    };
  }

  static fail(err) {
    return {
      type: 'SKILL_FETCH_FAILURE',
      payload: {
        err,
      },
    };
  }
}

export default s => Skill.export(s);
