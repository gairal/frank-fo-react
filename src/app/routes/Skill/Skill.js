import { connect } from 'react-redux';

import AbstractRoute from '../AbstractRoute';
import component from '../../components/Skill';

export default class Skill extends AbstractRoute {
  constructor(store) {
    super(store, 'skills', 'categories');
    this.initialState = {
      skills: [],
      isFetching: true,
    };

    this.component = component;
    this.mapDispatchToProps = {
      load: this.loadFactory(),
    };

    this.ACTION_HANDLERS = {
      [this.ACTIONS.FETCH]: state => ({
        ...state,
        isFetching: true,
      }),
      [this.ACTIONS.FETCH_SUCCESS]: (state, action) => ({
        ...state,
        skills: action.payload.skills,
        isFetching: false,
      }),
      [this.ACTIONS.FETCH_FAILURE]: state => ({
        ...state,
        isFetching: false,
      }),
    };
  }

  get connected() {
    return connect(state => ({
      skills: state.skills.skills,
      isFetching: state.skills.isFetching,
    }), this.mapDispatchToProps)(this.component);
  }

  static success(json) {
    return {
      type: 'FETCH_SUCCESS',
      payload: {
        skills: json,
      },
    };
  }

  loadFactory() {
    return () => AbstractRoute.load(this.API_URL, AbstractRoute.request, Skill.success);
  }
}
