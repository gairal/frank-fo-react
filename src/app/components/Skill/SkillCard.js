import React from 'react';
import PropTypes from 'prop-types';
import {
  Grid,
  Card,
  CardContent,
  ListItem,
  ListItemText,
  ListItemIcon,
  Typography,
} from '@material-ui/core';
import SkillLevel from './SkillLevel';

const SkillCard = ({ category }) => {
  const skillItem = skill => {
    let skillTitle = skill.name;
    if (skill.description) {
      skillTitle += ` | ${skill.description}`;
    }

    const progress = skill.level || 0;

    return (
      <Grid item xs={12} md={6} lg={4} key={skill.name}>
        <ListItem key={skill.name}>
          <ListItemIcon>
            <SkillLevel level={progress} />
          </ListItemIcon>
          <ListItemText
            primary={skillTitle}
            style={{
              flex: '1 1 0',
            }}
          />
        </ListItem>
      </Grid>
    );
  };

  return (
    <Card>
      <CardContent>
        <Typography variant="headline" component="h2">
          {category.name}
        </Typography>
        <Grid container>
          {category.skills ? category.skills.map(skillItem) : null}
        </Grid>
      </CardContent>
    </Card>
  );
};

SkillCard.propTypes = {
  category: PropTypes.shape().isRequired,
};

export default SkillCard;
