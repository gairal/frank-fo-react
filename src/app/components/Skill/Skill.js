import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Grid } from 'material-ui';
import Loader from '../core/Loader';
import SkillCard from './SkillCard';

export default class Skill extends Component {
  static propTypes = {
    load: PropTypes.func.isRequired,
    skills: PropTypes.arrayOf(PropTypes.object),
    isFetching: PropTypes.bool,
  }

  static defaultProps = {
    skills: [],
    isFetching: false,
  }

  componentDidMount() {
    this.props.load();
  }

  render() {
    return (
      <section>
        <Loader isFetching={this.props.isFetching} />
        <Grid container>
          {this.props.skills.map(category => (
            <Grid
              item
              xs={12}
              md={6}
              lg={4}
              key={category.name}
            >
              <SkillCard category={category} />
            </Grid>
          ))}
        </Grid>
      </section>
    );
  }
}
