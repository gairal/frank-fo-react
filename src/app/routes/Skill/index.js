import Skill from './Skill';

export default store => new Skill(store).route;
