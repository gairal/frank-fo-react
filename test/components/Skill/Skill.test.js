import React from 'react';
import renderer from 'react-test-renderer';
import Skill from '@/components/Skill';

describe('components > Skill', () => {
  it('Renders correctly default', () => {
    const component = renderer.create(
      <Skill load={jest.fn} skills={[{ name: 'test' }]} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders correctly with categories', () => {
    const component = renderer.create(
      <Skill
        load={jest.fn}
        skills={[
          {
            name: 'test',
            skills: [
              {
                name: 'test1',
                description: 'test',
                level: 77,
              },
              {
                name: 'test2',
              },
            ],
          },
        ]}
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
