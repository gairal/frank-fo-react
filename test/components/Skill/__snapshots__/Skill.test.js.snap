// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components > Skill Renders correctly default 1`] = `
<section>
  <div
    className="MuiGrid-container-1"
  >
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-40"
    >
      <div
        className="MuiPaper-root-98 MuiPaper-elevation2-102 MuiPaper-rounded-99 MuiCard-root-97"
      >
        <div
          className="MuiCardContent-root-125"
        >
          <h2
            className="MuiTypography-root-126 MuiTypography-headline-131"
          >
            test
          </h2>
          <div
            className="MuiGrid-container-1"
          />
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`components > Skill Renders correctly with categories 1`] = `
<section>
  <div
    className="MuiGrid-container-1"
  >
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-40"
    >
      <div
        className="MuiPaper-root-98 MuiPaper-elevation2-102 MuiPaper-rounded-99 MuiCard-root-97"
      >
        <div
          className="MuiCardContent-root-125"
        >
          <h2
            className="MuiTypography-root-126 MuiTypography-headline-131"
          >
            test
          </h2>
          <div
            className="MuiGrid-container-1"
          >
            <div
              className="MuiGrid-item-2 MuiGrid-grid-xs-12-40 MuiGrid-grid-md-6-62 MuiGrid-grid-lg-4-74"
            >
              <li
                className="MuiListItem-root-150 MuiListItem-default-153 MuiListItem-gutters-157"
                disabled={false}
              >
                <div
                  aria-valuenow={0}
                  className="MuiCircularProgress-root-161 MuiCircularProgress-colorPrimary-164 MuiCircularProgress-static-162"
                  role="progressbar"
                  style={
                    Object {
                      "height": 40,
                      "transform": "rotate(-90deg)",
                      "width": 40,
                    }
                  }
                >
                  <svg
                    className="MuiCircularProgress-svg-166"
                    viewBox="22 22 44 44"
                  >
                    <circle
                      className="MuiCircularProgress-circle-167 MuiCircularProgress-circleStatic-168"
                      cx={44}
                      cy={44}
                      fill="none"
                      r={19.5}
                      strokeWidth={5}
                      style={
                        Object {
                          "strokeDasharray": "122.522",
                          "strokeDashoffset": "122.522px",
                        }
                      }
                    />
                  </svg>
                </div>
                <div
                  className="MuiListItemText-root-170"
                  style={
                    Object {
                      "flex": "1 1 0",
                    }
                  }
                >
                  <span
                    className="MuiTypography-root-126 MuiTypography-subheading-133 MuiListItemText-primary-173"
                  >
                    test1 | test
                  </span>
                </div>
              </li>
            </div>
            <div
              className="MuiGrid-item-2 MuiGrid-grid-xs-12-40 MuiGrid-grid-md-6-62 MuiGrid-grid-lg-4-74"
            >
              <li
                className="MuiListItem-root-150 MuiListItem-default-153 MuiListItem-gutters-157"
                disabled={false}
              >
                <div
                  aria-valuenow={0}
                  className="MuiCircularProgress-root-161 MuiCircularProgress-colorPrimary-164 MuiCircularProgress-static-162"
                  role="progressbar"
                  style={
                    Object {
                      "height": 40,
                      "transform": "rotate(-90deg)",
                      "width": 40,
                    }
                  }
                >
                  <svg
                    className="MuiCircularProgress-svg-166"
                    viewBox="22 22 44 44"
                  >
                    <circle
                      className="MuiCircularProgress-circle-167 MuiCircularProgress-circleStatic-168"
                      cx={44}
                      cy={44}
                      fill="none"
                      r={19.5}
                      strokeWidth={5}
                      style={
                        Object {
                          "strokeDasharray": "122.522",
                          "strokeDashoffset": "122.522px",
                        }
                      }
                    />
                  </svg>
                </div>
                <div
                  className="MuiListItemText-root-170"
                  style={
                    Object {
                      "flex": "1 1 0",
                    }
                  }
                >
                  <span
                    className="MuiTypography-root-126 MuiTypography-subheading-133 MuiListItemText-primary-173"
                  >
                    test2
                  </span>
                </div>
              </li>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
